bedrock/base/i18n.py

```diff
@@ -0,0 +1,165 @@
+            list(urls),
+        )
+    ]
+
+
+def normalize_language(language):
+    """
+    Given a language code, returns the language code supported by Bedrock
+    in the proper case, for example "eN-us" --> "en-US" - or None if
+    Bedrock doesn't support the language code.
+
+    We also convert it to a specified fallback language if the actual
+    presented code is not available _and_ we have a fallback specced in
+    settings.
+
+    """
+
+    if not language:
+        return None
+
+    if settings.IS_POCKET_MODE:
+        lang_code = language.lower()
+    else:
+        lang_code = language
+
+    if lang_code in settings.LANGUAGE_URL_MAP_WITH_FALLBACKS:
+        return settings.LANGUAGE_URL_MAP_WITH_FALLBACKS[lang_code]
+
+    # Reformat the lang code to be mixed-case, as we expect
+    # them to be for our lookup
+    lang, _partition, classifier = language.partition("-")
+    lang = lang.lower()  # this part is _always_ lowercase
+
+    _special_cases = ["hant"]
+
+    if classifier:
+        if "-" in classifier or classifier in _special_cases:
+            _classifier, _partition, rest = classifier.partition("-")
+            if _classifier.lower() in _special_cases:
+                # support zh-hant-tw -> zh-Hant-TW
+                _classifier = capfirst(_classifier)
+                if rest:
+                    classifier = f"{_classifier}-{rest.upper()}"
+                else:
+                    classifier = _classifier
+            else:
+                # Support patterns like ja-JP-mac
+                classifier = f"{_classifier.upper()}-{rest}"
+        else:
+            classifier = classifier.upper()
+
+        lang_code = f"{lang}-{classifier}"
+    else:
+        lang_code = lang
+
+    try:
+        return settings.LANGUAGE_URL_MAP_WITH_FALLBACKS[lang_code]
+    except KeyError:
+        pre = lang_code.split("-")[0]
+        return settings.LANGUAGE_URL_MAP_WITH_FALLBACKS.get(pre)
+
+
+def split_path_and_polish_lang(path_):
+    """
+    Split the requested path into (lang, path) and
+    switches to a supported lang if available
+
+    locale will be empty if it isn't found.
+
+    Returns:
+        - lang code (str, may be empty)
+        - remaining URL path (str)
+        - whether or not the lang code changed (bool)
+
+    """
+    path = path_.lstrip("/")
+
+    # Use partition instead of split since it always returns 3 parts
+    extracted_lang, _, rest = path.partition("/")
+
+    supported_lang = normalize_language(extracted_lang)
+    different = extracted_lang != supported_lang
+
+    if extracted_lang and supported_lang:
+        return supported_lang, rest, different
+    else:
+        return "", path, False
+
+
+def get_language_from_headers(request):
+    """
+    Return a locale code we support on the site using the
+    user's Accept-Language header to determine which is best. This
+    mostly follows the RFCs but read bug 439568 for details.
+    """
```

Comment:
**note:**  (to self): reading this now, maybe we can just use Django's own helpers for this